File: backend/app/schemas/linked_entry.py

"""Linked entry schemas for API validation."""
from datetime import date as date_type, datetime as datetime_type
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.linked_entry import LinkType, LinkStatus


# LinkedEntry schemas

class LinkedEntryBase(BaseModel):
    """Base linked entry schema."""
    link_type: LinkType = Field(..., description="Type of link")
    counterparty_name: str = Field(..., min_length=1, max_length=200, description="Counterparty name")
    total_amount: Decimal = Field(..., gt=0, description="Total amount")
    user_amount: Optional[Decimal] = Field(default=None, gt=0, description="User's share (for split payments)")
    notes: Optional[str] = Field(default=None, max_length=1000, description="Optional notes")


class LinkedEntryCreate(BaseModel):
    """Schema for creating a linked entry."""
    primary_transaction_id: int = Field(..., gt=0, description="Primary transaction ID")
    link_type: LinkType = Field(..., description="Type of link")
    counterparty_name: str = Field(..., min_length=1, max_length=200, description="Counterparty name")
    user_amount: Optional[Decimal] = Field(default=None, gt=0, description="User's share (for split payments only)")
    notes: Optional[str] = Field(default=None, max_length=1000, description="Optional notes")


class LinkedEntryUpdate(BaseModel):
    """Schema for updating a linked entry."""
    counterparty_name: Optional[str] = Field(default=None, min_length=1, max_length=200)
    user_amount: Optional[Decimal] = Field(default=None, gt=0)
    notes: Optional[str] = Field(default=None, max_length=1000)


class MarkAsSplitRequest(BaseModel):
    """Schema to mark a transaction as split payment."""
    counterparty_name: str = Field(..., min_length=1, max_length=200)
    user_amount: Decimal = Field(..., gt=0, description="Your share of the cost")
    notes: Optional[str] = Field(default=None, max_length=1000)


class MarkAsLoanRequest(BaseModel):
    """Schema to mark a transaction as a loan."""
    counterparty_name: str = Field(..., min_length=1, max_length=200)
    notes: Optional[str] = Field(default=None, max_length=1000)


class MarkAsDebtRequest(BaseModel):
    """Schema to mark a transaction as a debt."""
    counterparty_name: str = Field(..., min_length=1, max_length=200)
    notes: Optional[str] = Field(default=None, max_length=1000)



# LinkedTransaction schemas

class LinkTransactionRequest(BaseModel):
    """Request to link a transaction to an entry."""
    transaction_id: int = Field(..., gt=0, description="Transaction ID to link")


class LinkedTransactionResponse(BaseModel):
    """Schema for linked transaction response."""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    linked_entry_id: int
    transaction_id: int
    amount: Decimal
    created_at: datetime_type
    date: Optional[date_type] = None
    description: Optional[str] = None



class LinkedEntryResponse(BaseModel):
    """Schema for linked entry response."""
    model_config = ConfigDict(from_attributes=True)
    
    id: int
    link_type: LinkType
    primary_transaction_id: int
    counterparty_name: str
    total_amount: Decimal
    user_amount: Optional[Decimal]
    pending_amount: Decimal
    status: LinkStatus
    notes: Optional[str]
    created_at: datetime_type
    updated_at: datetime_type
    linked_transactions: list[LinkedTransactionResponse] = []


class LinkedEntryWithDetails(LinkedEntryResponse):
    """Schema for linked entry with transaction details."""
    primary_transaction_description: Optional[str] = None
    primary_transaction_date: Optional[str] = None
    linked_transactions: list[LinkedTransactionResponse] = []
    settled_amount: Decimal = Field(default=Decimal("0.00"), description="Total amount settled")

File: SlaveOfCapitalism/SlaveOfCapitalism/Models/LinkedEntry.swift
import Foundation

struct LinkedEntryResponse: Codable, Identifiable, Sendable {
    let id: Int
    let linkType: LinkType
    let primaryTransactionId: Int
    let counterpartyName: String
    let totalAmount: Decimal
    let userAmount: Decimal?
    let pendingAmount: Decimal
    let status: LinkStatus
    let notes: String?
    let createdAt: String
    let updatedAt: String
    let linkedTransactions: [LinkedTransactionResponse]
}

struct LinkedEntryWithDetails: Codable, Identifiable, Sendable {
    let id: Int
    let linkType: LinkType
    let primaryTransactionId: Int
    let counterpartyName: String
    let totalAmount: Decimal
    let userAmount: Decimal?
    let pendingAmount: Decimal
    let status: LinkStatus
    let notes: String?
    let createdAt: String
    let updatedAt: String
    let linkedTransactions: [LinkedTransactionResponse]
    let primaryTransactionDescription: String?
    let primaryTransactionDate: String?
    let settledAmount: Decimal
}

struct LinkedTransactionResponse: Codable, Identifiable, Sendable {
    let id: Int
    let linkedEntryId: Int
    let transactionId: Int
    let amount: Decimal
    let createdAt: String
    let date: String?
    let description: String?
}

struct LinkedEntryCreate: Codable, Sendable {
    let primaryTransactionId: Int
    let linkType: LinkType
    let counterpartyName: String
    var userAmount: Decimal?
    var notes: String?
}

struct LinkedEntryUpdate: Codable, Sendable {
    var counterpartyName: String?
    var userAmount: Decimal?
    var notes: String?
}

struct LinkTransactionRequest: Codable, Sendable {
    let transactionId: Int
}

struct MarkAsSplitRequest: Codable, Sendable {
    let counterpartyName: String
    let userAmount: Decimal
    var notes: String?
}

struct MarkAsLoanRequest: Codable, Sendable {
    let counterpartyName: String
    var notes: String?
}

struct MarkAsDebtRequest: Codable, Sendable {
    let counterpartyName: String
    var notes: String?
}

struct OwedSummary: Codable, Sendable {
    let totalOwed: Decimal
    let pendingCount: Int
}

struct DebtSummary: Codable, Sendable {
    let totalDebt: Decimal
    let pendingCount: Int
}

enum LinkedEntryPresentation {
    static func ownerBadgeText(for transaction: TransactionWithDetails) -> String {
        if let linkType = transaction.linkedEntry?.linkType {
            return ownerBadgeText(for: linkType)
        }

        switch transaction.classification {
        case .splitPayment:
            return ownerBadgeText(for: .splitPayment)
        case .installment:
            return ownerBadgeText(for: .installment)
        default:
            return "Linked Entry"
        }
    }

    static func ownerBadgeText(for linkType: LinkType) -> String {
        switch linkType {
        case .splitPayment:
            return "Split Entry"
        case .loan:
            return "Loan Entry"
        case .debt:
            return "Debt Entry"
        case .installment:
            return "Installment Entry"
        }
    }

    static func linkActionTitle(for linkType: LinkType) -> String {
        switch linkType {
        case .splitPayment, .loan:
            return "Link Repayment"
        case .debt:
            return "Link Payment"
        case .installment:
            return "Link Charge"
        }
    }

    static func linkSheetTitle(for linkType: LinkType) -> String {
        linkActionTitle(for: linkType)
    }

    static func emptyStateDescription(for linkType: LinkType) -> String {
        switch linkType {
        case .splitPayment, .loan:
            return "Create the repayment transaction first, then link it here."
        case .debt:
            return "Create the payment transaction first, then link it here."
        case .installment:
            return "Create the matching charge transaction first, then link it here."
        }
    }

    static func linkedTransactionsLabel(for linkType: LinkType) -> String {
        switch linkType {
        case .splitPayment, .loan:
            return "Payments Received"
        case .debt:
            return "Repayments Made"
        case .installment:
            return "Charges Recorded"
        }
    }
}
